drivers/spi/spi_andes_atcspi200.c

```diff
@@ -162,31 +191,412 @@ static int configure(const struct device *dev,
 	return 0;
 }
 
-static void transfer_next_chunk(const struct device *dev)
+
+#ifdef CONFIG_ANDES_SPI_DMA_MODE
+
+static int spi_dma_tx_load(const struct device *dev);
+static int spi_dma_rx_load(const struct device *dev);
+
+static inline void spi_tx_dma_enable(const struct device *dev)
 {
-	struct spi_atcspi200_data * const data = dev->data;
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Enable TX DMA */
+	sys_set_bits(SPI_CTRL(cfg->base), CTRL_TX_DMA_EN_MSK);
+}
+
+static inline void spi_tx_dma_disable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Disable TX DMA */
+	sys_clear_bits(SPI_CTRL(cfg->base), CTRL_TX_DMA_EN_MSK);
+}
+
+static inline void spi_rx_dma_enable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Enable RX DMA */
+	sys_set_bits(SPI_CTRL(cfg->base), CTRL_RX_DMA_EN_MSK);
+}
+
+static inline void spi_rx_dma_disable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Disable RX DMA */
+	sys_clear_bits(SPI_CTRL(cfg->base), CTRL_RX_DMA_EN_MSK);
+}
+
+static int spi_dma_move_buffers(const struct device *dev)
+{
+	struct spi_atcspi200_data *data = dev->data;
 	struct spi_context *ctx = &data->ctx;
-	int error = 0;
+	uint32_t error = 0;
 
-	size_t chunk_len = spi_context_max_continuous_chunk(ctx);
+	data->dma_rx.dma_blk_cfg.next_block = NULL;
+	data->dma_tx.dma_blk_cfg.next_block = NULL;
 
-	if (chunk_len > 0) {
-		data->chunk_len = chunk_len;
-		error = spi_transfer(dev, chunk_len);
-		if (error == 0) {
+	if (spi_context_tx_on(ctx)) {
+		error = spi_dma_tx_load(dev);
+		if (error != 0) {
+			return error;
+		}
+	}
+
+	if (spi_context_rx_on(ctx)) {
+		error = spi_dma_rx_load(dev);
+		if (error != 0) {
+			return error;
+		}
+	}
+
+	return 0;
+}
+
+static inline void dma_rx_callback(const struct device *dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct spi_atcspi200_data *data = user_data;
```

Comment:
Why not use the device itself as the user_data for dma here instead if you need to get the struct device?

Can drop the self_dev pointer in data then